python/lsst/cp/pipe/bfCoeffCalc.py

```diff
@@ -0,0 +1,1272 @@
+        if the datasets already exist as this task exists to make them.
+        """
+        if self.datasetType is None:
+            raise RuntimeError("Must call setDatasetType first")
+        butler = namespace.butler
+        for dataId in self.idList:
+            refList = list(butler.subset(datasetType=self.datasetType, level=self.level, dataId=dataId))
+            # exclude nonexistent data
+            # this is a recursive test, e.g. for the sake of "raw" data
+            if not refList:
+                namespace.log.warn("No data found for dataId=%s", dataId)
+                continue
+            self.refList += refList
+
+
+class BfTask(pipeBase.CmdLineTask):
+    """Bright-fatter effect coefficient calculation task.
+
+    See http://ls.st/ldm-151 Chapter 4, Calibration Products Production for further details
+    regarding the inputs and outputs.
+    """
+
+    RunnerClass = BfTaskRunner
+    ConfigClass = BfTaskConfig
+    _DefaultName = "bf"
+
+    def __init__(self, *args, **kwargs):
+        """Constructor for the BfTask."""
+        pipeBase.CmdLineTask.__init__(self, *args, **kwargs)
+        self.makeSubtask("isr")
+
+        self.debug = lsstDebug.Info(__name__)
+        if self.debug.enabled:
+            self.log.info("Running with debug enabled...")
+            # if we're displaying, test it works, and save the displays for later
+            # it's worth testing here as displays are flaky and sometimes can't be contacted
+            # and given processing takes a while, it's a shame to fail late due to display issues
+            if self.debug.display:
+                try:
+                    afwDisp.setDefaultBackend(self.debug.displayBackend)
+                    afwDisp.Display.delAllDisplays()
+                    self.disp1 = afwDisp.Display(0, open=True)
+                    self.disp2 = afwDisp.Display(1, open=True)
+
+                    im = afwImage.ImageF(1, 1)
+                    im.array[:] = [[1]]
+                    self.disp1.mtv(im)
+                    self.disp1.erase()
+                except NameError:
+                    self.debug.display = False
+                    self.log.warn('Failed to setup/connect to display! Debug display has been disabled')
+
+        plt.interactive(False)  # stop windows popping up when plotting. When headless, use 'agg' backend too
+        self.config.validateIsrConfig(self.log)
+        self.config.validate()
+        self.config.freeze()
+
+    @classmethod
+    def _makeArgumentParser(cls):
+        """Augment argument parser for the BfTask."""
+        parser = pipeBase.ArgumentParser(name=cls._DefaultName)
+        parser.add_argument("--visitPairs", help="The list of visit pairs to use, as a list of tuples "
+                            "enclosed in quotes e.g. \"(123,456),(789,987),(654,321)\""
+                            "NB: must be comma-separated-tuples with no spaces, enclosed in quotes!")
```

Comment:
Oh. My. God. THIS! 🤯 This is the answer I was looking for all along! How did nobody else mention that I was reinventing the wheel rather than shouting at me about my lack of regex knowledge! 😄

Thank you, this is indeed the elegant (and pre-existing solution) I needed. `nargs="*"` for the win! 🎉

It looks to me like I still need a (now very slimmed down) taskRunner to pass through the argument though, is that right? It now just looks like

```
    @staticmethod
    def getTargetList(parsedCmd, **kwargs):
        """Pass the visit list through explicitly."""
        return pipeBase.TaskRunner.getTargetList(parsedCmd, visitPairs=parsedCmd.visitPairs, **kwargs)
```

Reply:
I had a little play seeing if I could get it through with `**kwargs` but failed. Could you just take a look at 16beda01764f3370818306624f2552d9feb93a7d and check that that is the best that can be done, _i.e._ I can't get rid of the TaskRunner altogether please?

Reply:
I tried passing them through un-parsed with `**kwargs` and doing the parsing later so that the whole TaskRunner could be done away with but couldn't get it working. Could you confirm that you think this is the best that can be done please?

Reply:
I think having a custom task runner is very reasonable because of the need to pass extra information. As to where to parse the visit pairs (converting them from a string `"123,456"` to a pair of integers `(123, 456)`): clearly it should be done before calling the task, which I think means it has to be done as part of parsing the arguments or in the task runner. Doing it while parsing the command optimal, if you can pull it off, as that catches any problems as soon as possible and allows you to print the usual "I could not parse this command, here is the command help" message. The argument parser does this for other arguments using a callback specified by the `action` argument to `add_argument`. See `class IdValueAction` in `argumentParser.py` for an example of a callback, and how it is used. There are a number of similar classes defined nearby.

Reply:
OK, glad I'm not doing it too wrong then. It seems like this one liner is very simple, and much more understandable (to me at least) than using callbacks etc, so as long as you don't think this needs moving elsewhere I'll just leave it as-is.